lib/runtime-core/src/tiering.rs

```diff
@@ -0,0 +1,236 @@
+use crate::backend::{Compiler, CompilerConfig};
+use crate::compile_with_config;
+use crate::fault::{
+    catch_unsafe_unwind, ensure_sighandler, pop_code_version, push_code_version, with_ctx,
+};
+use crate::fault::{set_wasm_interrupt_on_ctx, was_sigint_triggered_fault};
+use crate::import::ImportObject;
+use crate::instance::Instance;
+use crate::module::{Module, ModuleInfo};
+use crate::state::{x64::invoke_call_return_on_stack, CodeVersion, InstanceImage};
+use crate::vm::Ctx;
+
+use std::cell::Cell;
+use std::sync::{Arc, Mutex};
+
+struct Defer<F: FnOnce()>(Option<F>);
+impl<F: FnOnce()> Drop for Defer<F> {
+    fn drop(&mut self) {
+        if let Some(f) = self.0.take() {
+            f();
+        }
+    }
+}
+
+pub enum ShellExitOperation {
+    ContinueWith(InstanceImage),
+}
+
+pub struct InteractiveShellContext {
+    pub image: Option<InstanceImage>,
+    pub patched: bool,
+}
+
+struct OptimizationState {
+    outcome: Mutex<Option<OptimizationOutcome>>,
+}
+
+struct OptimizationOutcome {
+    module: Module,
+}
+
+#[repr(transparent)]
+struct CtxWrapper(*mut Ctx);
+unsafe impl Send for CtxWrapper {}
+unsafe impl Sync for CtxWrapper {}
+
+unsafe fn do_optimize(
+    binary: &[u8],
+    compiler: Box<dyn Compiler>,
+    ctx: &Mutex<CtxWrapper>,
+    state: &OptimizationState,
+) {
+    let module = match compile_with_config(
+        &binary[..],
+        &*compiler,
+        CompilerConfig {
+            symbol_map: None,
+            track_state: true,
+            ..Default::default()
```

Comment:
This doesn't work with the `--enable-simd` and `--enable-threads` flags, then? Maybe the `compiler` variable has a field with its CompilerConfig and you can use the fields from that?

Reply:
This is already checked in `wasmer.rs`.